File: csharp/engine/src/Loader.cs
using System.Text.Json;
using Ball.Shared;
using Ball.V1;
using Google.Protobuf;

namespace Ball.Engine;

/// <summary>
/// Loads a target Ball <see cref="Program"/> into the runtime (issue #383) — the
/// C# sibling of <c>rust/engine/src/loader.rs</c> / the TS wrapper's
/// <c>protoWrap</c>.
///
/// <para>Two entry points, mirroring the CLI's two input formats:</para>
/// <list type="bullet">
/// <item><see cref="ParseJson"/> — proto3-JSON <c>.ball.json</c> (the
/// examples/conformance format). Strips the cosmetic <c>@type</c> Any
/// envelope.</item>
/// <item><see cref="ParseBinary"/> — binary protobuf (a <c>google.protobuf.Any</c>
/// envelope wrapping the <c>ball.v1.Program</c>).</item>
/// </list>
///
/// <para>Both return the typed <see cref="Program"/> (for the wrapper's own
/// structural needs) <b>and</b> a canonical proto3-JSON <see cref="BallValue"/>
/// view of it — a tree of insertion-ordered <see cref="BallMap"/>s keyed by
/// camelCase <c>jsonName</c>s, oneofs represented by the set field being
/// present. That view is exactly the shape the compiled self-hosted engine reads
/// through the <see cref="BallProto"/> access-pattern functions
/// (<c>whichExpr</c>/<c>hasBody</c>/…).</para>
/// </summary>
public static class Loader
{
    /// <summary>
    /// Parse a proto3-JSON <c>.ball.json</c> program (stripping the cosmetic
    /// <c>@type</c> Any envelope) into a typed <see cref="Program"/> plus its
    /// canonical <see cref="BallValue"/> view.
    /// </summary>
    public static (Program Program, BallValue View) ParseJson(string json)
    {
        string body;
        try
        {
            body = StripTypeEnvelope(json);
        }
        catch (JsonException e)
        {
            throw new EngineException($"parse error: {e.Message}");
        }

        Program program;
        try
        {
            var parser = new JsonParser(JsonParser.Settings.Default.WithIgnoreUnknownFields(true));
            program = parser.Parse<Program>(body);
        }
        catch (InvalidProtocolBufferException e)
        {
            throw new EngineException($"not a valid ball.v1.Program: {e.Message}");
        }

        return (program, BuildView(program));
    }

    /// <summary>
    /// Parse a binary-protobuf program — a <c>google.protobuf.Any</c> envelope
    /// wrapping the <c>ball.v1.Program</c> — into a typed <see cref="Program"/>
    /// plus its canonical <see cref="BallValue"/> view.
    /// </summary>
    public static (Program Program, BallValue View) ParseBinary(byte[] bytes)
    {
        Program program;
        try
        {
            var any = Google.Protobuf.WellKnownTypes.Any.Parser.ParseFrom(bytes);
            program = any.Unpack<Program>();
        }
        catch (InvalidProtocolBufferException e)
        {
            throw new EngineException($"not a valid ball.v1.Program: {e.Message}");
        }

        return (program, BuildView(program));
    }

    /// <summary>Strip the top-level <c>@type</c> Any-envelope key, returning the remaining JSON body.</summary>
    private static string StripTypeEnvelope(string json)
    {
        using var doc = JsonDocument.Parse(json);
        if (doc.RootElement.ValueKind != JsonValueKind.Object)
        {
            return json;
        }

        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(stream))
        {
            writer.WriteStartObject();
            foreach (var property in doc.RootElement.EnumerateObject())
            {
                if (property.NameEquals("@type"))
                {
                    continue;
                }

                property.WriteTo(writer);
            }

            writer.WriteEndObject();
        }

        return System.Text.Encoding.UTF8.GetString(stream.ToArray());
    }

    /// <summary>
    /// The canonical proto3-JSON <see cref="BallValue"/> view of
    /// <paramref name="program"/>. Serializes with <c>FormatDefaultValues</c> so
    /// every proto3 default is materialized (an absent repeated field becomes
    /// <c>[]</c>, an absent string <c>""</c>) — matching how the reference engine
    /// reads a program (its protobuf getters always return a field's default,
    /// never null) — then reconstructs the raw <c>google.protobuf.Struct</c>
    /// shape for every <c>metadata</c> field (the engine reads metadata through
    /// the proto <c>Value</c> API it was authored against).
    /// </summary>
    public static BallValue BuildView(Program program)
    {
        var formatter = new JsonFormatter(JsonFormatter.Settings.Default.WithFormatDefaultValues(true));
        var json = formatter.Format(program);
        using var doc = JsonDocument.Parse(json);
        return NormalizeMetadata(JsonToBallValue(doc.RootElement));
    }

    /// <summary>
    /// Convert a proto3-JSON element into a <see cref="BallValue"/> tree. Objects
    /// become insertion-ordered <see cref="BallMap"/>s; arrays become
    /// <see cref="BallList"/>s; an integral number becomes an <see cref="BallInt"/>,
    /// else a <see cref="BallDouble"/>. Proto3-JSON int64 fields arrive as
    /// <em>strings</em> (left as-is — the engine's literal path parses them), and
    /// a <c>bytesValue</c> base64 string is decoded to raw bytes.
    /// </summary>
    public static BallValue JsonToBallValue(JsonElement element) => element.ValueKind switch
    {
        JsonValueKind.Null => BallValue.Null,
        JsonValueKind.True => BallValue.Bool(true),
        JsonValueKind.False => BallValue.Bool(false),
        JsonValueKind.String => BallValue.Str(element.GetString()!),
        JsonValueKind.Number => element.TryGetInt64(out var i)
            ? BallValue.Int(i)
            : BallValue.Double(element.GetDouble()),
        JsonValueKind.Array => BuildList(element),
        JsonValueKind.Object => BuildMap(element),
        _ => BallValue.Null,
    };

    private static BallValue BuildList(JsonElement array)
    {
        var list = new BallList();
        foreach (var item in array.EnumerateArray())
        {
            list.Add(JsonToBallValue(item));
        }

        return list;
    }

    private static BallValue BuildMap(JsonElement obj)
    {
        var map = new BallMap();
        foreach (var property in obj.EnumerateObject())
        {
            // A Literal.bytesValue serializes as base64; the engine reads a bytes
            // literal as a list of byte ints, so decode it here.
            if (property.NameEquals("bytesValue") && property.Value.ValueKind == JsonValueKind.String)
            {
                map.Set("bytesValue", BallValue.Bytes(Convert.FromBase64String(property.Value.GetString()!)));
                continue;
            }

            // A Literal.doubleValue is a proto double, but proto3-JSON renders a
            // whole double (`9.0`) as a bare integer (`9`) — which the generic
            // number path below would load as a BallInt, dropping the double-ness
            // the engine's BallDouble literal path (and its trailing `.0`
            // formatting) depends on. Coerce it to a BallDouble regardless of the
            // JSON token's shape (a fractional value already loads as a double).
            if (property.NameEquals("doubleValue") && property.Value.ValueKind == JsonValueKind.Number)
            {
                map.Set("doubleValue", BallValue.Double(property.Value.GetDouble()));
                continue;
            }

            map.Set(property.Name, JsonToBallValue(property.Value));
        }

        return map;
    }

    // ════════════════════════════════════════════════════════════
    // metadata Struct reconstruction (issue #300 / rust loader parity)
    // ════════════════════════════════════════════════════════════

    /// <summary>
    /// Reconstruct the raw <c>google.protobuf.Struct</c> proto shape for every
    /// <c>metadata</c> field of the program view. Proto3-JSON collapses a Struct
    /// to a plain object (<c>{kind: "function", …}</c>), but the self-hosted
    /// engine reads metadata through the proto object API
    /// (<c>func.metadata.fields['kind'].stringValue</c>, the <c>whichKind</c>/
    /// <c>hasStringValue</c> discriminators on each <c>Value</c>). Each metadata
    /// Struct is therefore re-expanded to <c>{fields: {key: Value}}</c>.
    /// </summary>
    private static BallValue NormalizeMetadata(BallValue value)
    {
        switch (value)
        {
            case BallMap map:
                var outMap = new BallMap(map.Count);
                foreach (var (key, val) in map.Entries())
                {
                    if (key == "metadata" && val is BallMap structMap)
                    {
                        outMap.Set(key, WrapStruct(structMap));
                    }
                    else
                    {
                        outMap.Set(key, NormalizeMetadata(val));
                    }
                }

                return outMap;
            case BallList list:
                var outList = new BallList();
                foreach (var item in list.Snapshot())
                {
                    outList.Add(NormalizeMetadata(item));
                }

                return outList;
            default:
                return value;
        }
    }

    /// <summary>A collapsed metadata object → the raw Struct shape <c>{fields: {key: Value}}</c>.</summary>
    private static BallValue WrapStruct(BallMap map)
    {
        var fields = new BallMap(map.Count);
        foreach (var (key, val) in map.Entries())
        {
            fields.Set(key, WrapValue(val));
        }

        var outMap = new BallMap(1);
        outMap.Set("fields", fields);
        return outMap;
    }

    /// <summary>A plain metadata value → a <c>google.protobuf.Value</c> wrapper (the arm keyed by its kind).</summary>
    private static BallValue WrapValue(BallValue value)
    {
        var outMap = new BallMap(1);
        switch (value)
        {
            case BallNull:
                outMap.Set("nullValue", BallValue.Int(0));
                break;
            case BallBool b:
                outMap.Set("boolValue", b);
                break;
            case BallInt i:
                outMap.Set("numberValue", BallValue.Double(i.Value));
                break;
            case BallDouble d:
                outMap.Set("numberValue", d);
                break;
            case BallString s:
                outMap.Set("stringValue", s);
                break;
            case BallList list:
                var values = new BallList();
                foreach (var item in list.Snapshot())
                {
                    values.Add(WrapValue(item));
                }

                var listValue = new BallMap(1);
                listValue.Set("values", values);
                outMap.Set("listValue", listValue);
                break;
            case BallMap m:
                outMap.Set("structValue", WrapStruct(m));
                break;
            default:
                outMap.Set("stringValue", BallValue.Str(value.ToString() ?? string.Empty));
                break;
        }

        return outMap;
    }
}

File: go/engine/loader.go
package engine

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strconv"

	ballrt "github.com/ball-lang/ball/go/runtime"
	ballv1 "github.com/ball-lang/ball/go/shared/gen/ball/v1"
	"google.golang.org/protobuf/encoding/protojson"
)

// Loader turns a typed Ball Program into the canonical proto3-JSON BallValue
// view the compiled self-hosted engine reads (epic #426 Phase 4) — the Go
// sibling of csharp/engine/src/Loader.cs and rust/engine/src/loader.rs.
//
// The view is a tree of insertion-ordered *ballrt.Map keyed by camelCase
// jsonNames, oneofs represented by which variant key is present, that the engine
// walks through the ball_proto access-pattern functions (whichExpr/hasBody/…).

// buildView serializes program with proto3 default values materialized (an
// absent repeated field becomes [], an absent string ""), then reconstructs the
// BallValue tree with the loader special-cases: a bytesValue base64 string is
// decoded to raw bytes, a doubleValue is forced to a Ball double, and each
// metadata Struct is re-expanded to the raw {fields:{key:Value}} proto shape.
func buildView(program *ballv1.Program) (ballrt.Value, error) {
	data, err := protojson.MarshalOptions{
		EmitDefaultValues: true,
		UseProtoNames:     false,
	}.Marshal(program)
	if err != nil {
		return nil, fmt.Errorf("serialize program view: %w", err)
	}

	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	value, err := decodeValue(dec, "")
	if err != nil {
		return nil, fmt.Errorf("decode program view: %w", err)
	}
	return normalizeMetadata(value), nil
}

// decodeValue reads one JSON value from dec, preserving object key order. key is
// the object key this value is stored under (for the bytesValue/doubleValue
// special-cases), or "" at the root / inside an array.
func decodeValue(dec *json.Decoder, key string) (ballrt.Value, error) {
	tok, err := dec.Token()
	if err != nil {
		return nil, err
	}
	switch t := tok.(type) {
	case json.Delim:
		switch t {
		case '{':
			return decodeObject(dec)
		case '[':
			return decodeArray(dec)
		}
		return nil, fmt.Errorf("unexpected delimiter %q", t)
	case string:
		if key == "bytesValue" {
			raw, err := base64.StdEncoding.DecodeString(t)
			if err != nil {
				return nil, fmt.Errorf("decode bytesValue: %w", err)
			}
			return raw, nil
		}
		return t, nil
	case json.Number:
		return decodeNumber(t, key)
	case bool:
		return t, nil
	case nil:
		return nil, nil
	}
	return nil, fmt.Errorf("unexpected token %T", tok)
}

func decodeObject(dec *json.Decoder) (ballrt.Value, error) {
	m := ballrt.NewMap()
	for dec.More() {
		keyTok, err := dec.Token()
		if err != nil {
			return nil, err
		}
		key, ok := keyTok.(string)
		if !ok {
			return nil, fmt.Errorf("object key not a string: %T", keyTok)
		}
		val, err := decodeValue(dec, key)
		if err != nil {
			return nil, err
		}
		m.Set(key, val)
	}
	// Consume the closing '}'.
	if _, err := dec.Token(); err != nil {
		return nil, err
	}
	return m, nil
}

func decodeArray(dec *json.Decoder) (ballrt.Value, error) {
	l := ballrt.NewList()
	for dec.More() {
		val, err := decodeValue(dec, "")
		if err != nil {
			return nil, err
		}
		l.Add(val)
	}
	// Consume the closing ']'.
	if _, err := dec.Token(); err != nil {
		return nil, err
	}
	return l, nil
}

// decodeNumber turns a JSON number into a Ball int or double. A doubleValue
// literal is always a double, even when proto3-JSON renders a whole double (9.0)
// as a bare integer (9); every other integral number is a Ball int.
func decodeNumber(n json.Number, key string) (ballrt.Value, error) {
	if key == "doubleValue" {
		f, err := n.Float64()
		if err != nil {
			return nil, err
		}
		return f, nil
	}
	if i, err := strconv.ParseInt(n.String(), 10, 64); err == nil {
		return i, nil
	}
	f, err := n.Float64()
	if err != nil {
		return nil, err
	}
	return f, nil
}

// normalizeMetadata reconstructs the raw google.protobuf.Struct proto shape for
// every metadata field (proto3-JSON collapses a Struct to a plain object, but
// the engine reads metadata through the proto object API).
func normalizeMetadata(value ballrt.Value) ballrt.Value {
	switch v := value.(type) {
	case *ballrt.Map:
		out := ballrt.NewMap()
		for _, key := range v.Keys() {
			val, _ := v.Get(key)
			if key == "metadata" {
				if structMap, ok := val.(*ballrt.Map); ok {
					out.Set(key, wrapStruct(structMap))
					continue
				}
			}
			out.Set(key, normalizeMetadata(val))
		}
		return out
	case *ballrt.List:
		out := ballrt.NewList()
		for _, item := range v.Items {
			out.Add(normalizeMetadata(item))
		}
		return out
	default:
		return value
	}
}

// wrapStruct turns a collapsed metadata object into the raw Struct shape
// {fields:{key:Value}}.
func wrapStruct(m *ballrt.Map) ballrt.Value {
	fields := ballrt.NewMap()
	for _, key := range m.Keys() {
		val, _ := m.Get(key)
		fields.Set(key, wrapValue(val))
	}
	out := ballrt.NewMap()
	out.Set("fields", fields)
	return out
}

// wrapValue turns a plain metadata value into a google.protobuf.Value wrapper
// (the arm keyed by its kind).
func wrapValue(value ballrt.Value) ballrt.Value {
	out := ballrt.NewMap()
	switch v := value.(type) {
	case nil:
		out.Set("nullValue", int64(0))
	case bool:
		out.Set("boolValue", v)
	case int64:
		out.Set("numberValue", float64(v))
	case float64:
		out.Set("numberValue", v)
	case string:
		out.Set("stringValue", v)
	case *ballrt.List:
		values := ballrt.NewList()
		for _, item := range v.Items {
			values.Add(wrapValue(item))
		}
		listValue := ballrt.NewMap()
		listValue.Set("values", values)
		out.Set("listValue", listValue)
	case *ballrt.Map:
		out.Set("structValue", wrapStruct(v))
	default:
		out.Set("stringValue", ballrt.ToStr(v))
	}
	return out
}
